### ISSUE TITLE
No workaround for advanced indexing

### ISSUE THREAD

**Issue by blythed** [pytorch/pytorch#2469]
I'm trying to do some complex setting of a slice of an intermediate output.
Ideally I want something like:
```python
x[:, ix, :] = y
```
However, pytorch only seems to allow subsetting in the first dimension:
The following is fine:
```python
x = torch.autograd.Variable(torch.randn(4, 2, 3))
ix = torch.LongTensor([2, 3]) 
x[ix] = y
```
However when I try and use transposes to get around the slice on the second dimension, this doesn't 
work:

```python
x = torch.autograd.Variable(torch.randn(4, 2, 3))
ix = torch.LongTensor([2, 3]) 
x = x.transpose(1, 0)
x[ix] = y
x = x.transpose(0, 1)
```
Then I get the error message:
```
RuntimeError: in-place operations can be only used on variables that don't share storage with any other variables, but detected that there are 2 objects sharing it
```
I have tried all combinations of copying variables etc. but I cannot see a way round how to assign this subslice. Is this a known issue?

**Comment by killeent** [comment]
@blythed what are you trying to do exactly? Given your top line:

```
x[:, ix, :] = y
```

Can you provide example `x`, `ix`, and `y` and what you expect the effect to be? It doesn't follow exactly from your subsequent code what you want.

**Comment by blythed** [comment]
Hi killeent, 

Thanks for your reply. An example:

```python
x = torch.randn(3, 4, 5)
y = torch.randn(3, 2, 5)
ix = torch.LongTensor([1, 3])
x[:, ix, :] = y
```
```x``` and ```y``` are tensors agreeing in the first and last dimensions. ```ix``` is a LongTensor 
referencing some subset of the second dimension of ```x```. I would like to set this subset with ```y```
as one might do in numpy style code.

Thanks,

Duncan

**Comment by killeent** [comment]
@blythed 

The code snippet you posted does indeed work in PyTorch. Can you clarify further want is going wrong for you?

**Comment by fmassa** [comment]
@killeent I think he meant when running his snippet using Variables (instead of tensors)

**Comment by blythed** [comment]
The error message for version 0.1.12_2 was : 

TypeError: indexing a tensor with an object of type torch.LongTensor. The only supported types are integers, slices, numpy scalars and torch.LongTensor or torch.ByteTensor as the only argument.

I just updated the version to 0.2.0_3.
Now there is a new problem as fmassa noticed:

``` python 
x = Variable(torch.randn(3, 4, 5))
y = Variable(torch.randn(3, 2, 5))
ix = Variable(torch.LongTensor([1, 3]))
x[:, ix, :] = y
```

This gives the error:

IndexError: When performing advanced indexing the indexing objects must be LongTensors or convertible to LongTensors.

Aren't ```Variable``` objects containing ```LongTensor``` objects convertible?

**Comment by soumith** [comment]
this is now fixed in master with the merge of #2587
